### DISCUSSION TITLE
How to update from 2020 version to 2022 version?

### DISCUSSION THREAD

**Discussion by glow12121** [datenstrom/community#765] [Ask a question] [answered]
Hi,

Today I wanted to edit my website after not touching it for several months. I updated Yellow at the command line and tried to build a static website. But I got this error:

```
$ php yellow.php build
ERROR building files: Please configure CoreStaticUrl in file 'system/extensions/yellow-system.ini'!
Yellow build: 0 files, 1 error
```

There is no `yellow-system.ini` in the folder `system/extensions/`. For this reason I  tried to make a new installation. Downloading Yellow master files, extensions and a theme, then copying my previous content and customized files to the `yellow-master` folder.  As a result, I got these errors:

```
$ php yellow.php serve
Datenstrom Yellow requires complete upload!
```

```
$ php yellow.php build
Datenstrom Yellow requires complete upload!
```
Tip: You can edit your own questions here on GitHub, click on `...` -> `Edit`.

**Comment by glow12121** [comment]
It seems that an updated Yellow is incompatible with the previous version.

**Reply by schulle4u** [reply]
Dit your installation work after copying the fresh files to your server? If not, please check [our troubleshooting gguide](https://datenstrom.se/yellow/help/troubleshooting#problems-during-installation) for problems during installation. 

> Copy all of the supplied files to the web server. Check if your FTP software shows an error message during upload. It sometimes happens that the data transfer was interrupted, then files are missing on the web server. After the missing installation files have been copied to the web server, the problem should be resolved.

**Comment by glow12121** [comment]
Thank you for the answer. But I always run Yellow locally, on my PC.

**Reply by IngoBelka** [reply]
What about to install a fresh system with all extensions you did use for your old system? After that you copy only the content folder into the fresh installation and look, what comes out then. I'd do this way.

**Comment by glow12121** [comment]
Thank you for the answer. So, I've tried doing this step by step.

1. Downloading and unpacking the fresh yellow-master.zip
2. Locally running `php yellow.php update` - **OK**
3. Downloading the extensions I use to the respective folder and running `php yellow.php update` - **OK**
4. Copying my old `content` folder to the new `yellow-master` folder - **SAME `Datenstrom Yellow requires complete upload!` MISTAKE**
5. I tried copying the `content` folder without `shared` subfolder - **SAME MISTAKE**

There must be some incompatibility in `content`.

I think it is the third time in the last couple of years that Yellow does not work after an update.

**Reply by schulle4u** [reply]
I think you are running the installation in wrong order. 

1. Download and install latest Yellow.
2. From command line, run `php yellow.php serve` and open a browser to complete the installation. 
3. Now install your desired extensions, and copy your old content files. 
4. Press CTRL+C to close the web server and run `php yellow.php update` to see if your website is working. 

HTH
Steffen

**Comment by IngoBelka** [comment]
If you got a complete fresh installation with your extensions, take a look into the different folders in content. Compare the content files of the fresh - there are only less ones - and the old system. I'm not shure, but are the files all *. md and not *. txt or something else? Eventually copy your old content step by step, until you get an error message. And use the builtin webserver as mentiond by @schulle4u .

**Comment by markseuffert** [comment]
Hello guys. Let's try another attempt in solving the problem. 😀

Can we go back a few steps and see what happened? It would be good if you had more information to narrow down the cause of the problem. It's not clear to me at this point what happened. From the error message it sounds like the website was an older version from before October 2020. That was when the system settings have been moved to another subfolder. I could be wrong. In most cases the updater takes care of such changes, but I have to admit that the updater is not perfect, it only works smoothly if one updates frequently. Skipping over multiple releases is not really recommended, because the updater doesn't carry around multiple years of backwards-compatibility. So I suggest to clarify the following: 

How often do you update your website, when was the last time? Can you show your log file? Which subfolders do you have in your system folder at the moment?  Do you have any backups of your website, so you can roll back to a previously working website? Do you have any additional information or questions regarding updates?

PS: If you hit reply, instead of answer, your GitHub comments appear in the same thread.

**Reply by glow12121** [reply]
> If you hit reply, instead of answer, your GitHub comments appear in the same thread.

I see.

Do you think I could email the website, in the raw state, to you so you take a look? It is around 30 mbs. Thanks anyway.

As to how often I have been updating it - I remember I ran `update` almost every time I used Yellow, but I cannot be certain of course. These are the dates when the three backup copies I kept were created. I created a backup *almost* every time I made considerable changes.

![Screenshot from 2022-03-22 18-20-13](https://user-images.githubusercontent.com/29035180/159516984-3ca31f47-d634-4f3c-a900-5310162699c9.png)

**Reply by markseuffert** [reply]
Can we investigate the problem here and clarify what happened?

**Reply by glow12121** [reply]
I see, let's try to figure it out here first.

* How often do you update your website, when was the last time?

![Screenshot from 2022-03-22 18-20-13](https://user-images.githubusercontent.com/29035180/159520996-56b67881-d350-40b5-af27-612733e2ecb5.png)

* Which subfolders do you have in your system folder at the moment?

The latest non-working iteration (after reinstallation attempt): extensions, layouts, themes, trash

The last working iteration: extensions, layouts, settings, themes, trash

* Do you have any backups of your website, so you can roll back to a previously working website?

Yes, I have three backups left.

* Do you have any additional information or questions regarding updates?

Can they possibly be performed automatically, in particular when another command is entered?

* Can you show your log file?

Here is the log of the latest iteration that does not work properly:

```
2022-03-17 18:20:59 info Install Datenstrom Yellow 0.8.19, PHP 7.4.27, Built-in 7.4.27, Linux
2022-03-17 18:21:07 info Install extension 'English 0.8.29'
2022-03-17 18:21:07 info Install extension 'Russian 0.8.29'
2022-03-17 18:21:07 info Install extension 'German 0.8.29'
2022-03-17 18:21:59 info Add user 'User'
2022-03-17 18:23:08 info Install extension 'Audio 0.8.9'
2022-03-17 18:23:08 info Install extension 'Berlin 0.8.11'
2022-03-17 18:23:08 info Install extension 'Breadcrumb 0.8.6'
2022-03-17 18:23:08 info Install extension 'Bundle 0.8.24'
2022-03-17 18:23:08 info Install extension 'Fontawesome 0.8.12'
2022-03-17 18:23:08 info Install extension 'Image 0.8.13'
2022-03-17 18:23:08 info Install extension 'Serve 0.8.17'
2022-03-17 18:23:08 info Install extension 'Slider 0.8.15'
2022-03-17 18:23:08 info Install extension 'Youtube 0.8.5'
```

Here is the log of the last iteration that worked:

```
2020-07-29 16:54:27 info Datenstrom Yellow 0.8.14, PHP 7.3.19, CLI 7.3.19, Linux
2020-07-29 16:54:27 info Install extension 'English 0.8.21'
2020-07-29 16:54:27 info Install extension 'French 0.8.21'
2020-07-29 16:54:27 info Install extension 'German 0.8.21'
2020-07-29 16:58:00 info Install extension 'Berlin 0.8.9'
2020-07-29 16:58:00 info Install extension 'Image 0.8.9'
2020-07-29 16:58:00 info Install extension 'Slider 0.8.5'
2020-07-29 16:58:00 info Install extension 'Minishare 0.8.3'
2020-07-29 16:58:00 info Install extension 'Youtube 0.8.4'
2020-07-29 17:52:20 info Install extension 'Fontawesome 0.8.6'
2020-10-09 15:24:35 info Update extension 'Command 0.8.22'
2020-10-09 15:24:35 info Update extension 'Core 0.8.19'
2020-10-09 15:24:35 info Update extension 'Edit 0.8.35'
2020-10-09 15:24:35 info Update extension 'Update 0.8.31'
2020-10-09 15:31:36 info Update extension 'Bundle 0.8.16'
2020-10-09 15:31:36 info Update extension 'Command 0.8.23'
2020-10-09 15:31:36 info Update extension 'Core 0.8.21'
2020-10-09 15:31:36 info Update extension 'English 0.8.24'
2020-10-09 15:31:36 info Update extension 'French 0.8.24'
2020-10-09 15:31:36 info Update extension 'German 0.8.24'
2020-10-09 15:31:36 info Update extension 'Image 0.8.10'
2020-10-09 15:31:36 info Update extension 'Slider 0.8.6'
2020-10-09 15:31:36 info Update extension 'Update 0.8.34'
2020-10-13 14:36:46 info Update extension 'Core 0.8.22'
2020-10-13 14:36:46 info Update extension 'Fontawesome 0.8.7'
2020-10-13 14:36:46 info Update extension 'Markdown 0.8.16'
2022-03-11 21:20:39 info Update extension 'Berlin 0.8.11'
2022-03-11 21:20:39 info Update extension 'Bundle 0.8.24'
2022-03-11 21:20:39 info Update extension 'Command 0.8.33'
2022-03-11 21:20:39 info Update extension 'Core 0.8.59'
2022-03-11 21:20:39 info Update extension 'Edit 0.8.56'
2022-03-11 21:20:39 info Update extension 'English 0.8.29'
2022-03-11 21:20:39 info Update extension 'Fontawesome 0.8.12'
2022-03-11 21:20:39 info Update extension 'French 0.8.29'
2022-03-11 21:20:39 info Update extension 'German 0.8.29'
2022-03-11 21:20:39 info Update extension 'Image 0.8.13'
2022-03-11 21:20:39 info Update extension 'Markdown 0.8.19'
2022-03-11 21:20:39 info Update extension 'Slider 0.8.15'
2022-03-11 21:20:39 info Update extension 'Stockholm 0.8.11'
2022-03-11 21:20:39 info Update extension 'Update 0.8.65'
2022-03-11 21:20:39 info Update extension 'Youtube 0.8.5'
```

**Reply by markseuffert** [reply]
Thank you for the detailed information. Let's see where we are standing. According to the log file, your website was running an October 2020 version which was updated March 2022. Unfortunately the [automatic update](https://github.com/annaesvensson/yellow-update) doesn't work with such an old version. There were some bigger API changes in the meantime. Here are some alternatives:

* You can transfer your website to a new installation of Datenstrom Yellow. Some assembly required. As Steffen wrote above the only "trick to remember" is to complete the installation, before transferring your content and media folders. After that it's customising your website as usual and updating your website once a year.
* You can roll back your website to the backup from December 2021. Do not install updates. As Steffen wrote in another discussion you are not forced to install updates, one can keep an old version indefinitely. It's not recommended, as we want that people use the latest extensions and become active in our community.
*  You can improve the update extension. As mentioned the updater doesn't carry around multiple years of backwards-compatibility. But this is a limitation of the current technology, not something written in stone, we might want to develop a better update mechanism.
* Alternatively contact the commercial support. You can get your entire website updated for a small fee.

To answer your question. No, updates are never performed automatically when another command is entered. Updates are only performed, if someone manually runs `php yellow.php update` at the command line or if someone manually clicks the `Update` button in the web browser. The "automatic update" refers to automatically converting old files to the new API. 

I am really sorry for the trouble the update has caused you, I agree it's not perfect.

**Reply by glow12121** [reply]
> I am really sorry for the trouble the update has caused you, I agree it's not perfect.

No, many thanks for the answer. I will see if I can handle transfering my stuff to the version. If not, I will probably address your support service.

I will write back here if I have more questions if you don't mind.

**Comment by glow12121** [comment]
Should I ask this question here or in a new thread? Could it be that in the newer version of Yellow the .footer has a different CSS formatting from the body? For some reason, the footer has a different style for links and generic text.

**Reply by markseuffert** [reply]
I am not aware of such changes. In the default layout/theme, the footer is a child element of the body, it inherits the style for links and text. Maybe check the DOM tree in your web bowser?

**Reply by glow12121** [reply]
I copied my old footer.md to content/shared - in the new Yellow there  was no footer.md in this folder. Is it right?

footer.md starts as follows:

```
---
Title: Footer
Status: shared
---
```

With a modified Berlin theme, somehow only some style elements apply to the footer. `text-align: center;` works but the text is grayish and hyperlinks look different from the rest of the page.

```
/* Footer */

.footer {
    margin-top: 2em;
    margin-bottom: 2em;
    text-align: center;
}
.footer .siteinfo {
    color: datenstrom/yellow#888;
    padding-top: 1em;
    padding-bottom: 1em;
}
.footer .siteinfo a {
    color: #555;
    text-decoration: none;
}
.footer .siteinfo a:hover {
    color: #555;
    text-decoration: underline;
}
```

If I copy what is below from `body` style to `.footer` style, the footer font becomes bigger than the rest, but  still remains gray.

```
    color: #555;
    font-family: sans-serif;
    font-size: 1.1em;
```

**Reply by markseuffert** [reply]
Yes, a brand new installation doesn't come with a Markdown footer. Here's an example:

```
---
Title: Footer
Status: shared
---
Copyright Datenstrom. All rights reserved. &nbsp;&nbsp;&nbsp; 
[Privacy policy](/privacy/) | [Terms of use](/terms/) | [Sitemap](/sitemap/)
```

About greyish footer colours, I think this how the [Berlin theme](https://github.com/annaesvensson/yellow-berlin) was designed. Some themes use different link colours for different parts of the screen. Feel free to customise everything, use your CSS powers. 😁

**Reply by glow12121** [reply]
I got it -- `.footer .siteinfo` is what was responsible for the whole footer's decoration.

**Reply by markseuffert** [reply]
I also found it confusing. I updated the footer colours in the Berlin theme, now available in latest version.

**Comment by markseuffert** [comment, marked as answer]
It looks like the problem is solved, thanks @schulle4u and @IngoBelka.

Broken update -> manually transfer your website to a new installation, better update mechanism is needed.

New installation ->  complete the installation before doing anything else, better error messages are needed.

![Wow, a different error message, finally some progress](https://user-images.githubusercontent.com/2429165/160257032-08594385-2ff8-4754-b3f5-89e708f9cd15.jpg)

**Reply by glow12121** [reply]
> better update mechanism is needed

Maybe this repeatition is another symptom of this:

```
user@pc ~/Stuff/Yellow/yellow-master $ php yellow.php update all
Your website has been updated: Please run command again
user@pc ~/Stuff/Yellow/yellow-master $ php yellow.php update all
Berlin 0.8.12
Yellow update: Website updated, 1 update installed
```

**Reply by markseuffert** [reply]
Done... update mechanism and error messages have been improved.

Tested with Datenstrom Yellow 0.8.15 - 0.8.19, August 2020 and later.

**Comment by glow12121** [comment]
Thank you, that's good news.
